Block Styles: Extend block style variations as mechanism for achieving section styling

aaronrobertshaw opened the pull request:
Provide users with the ability to style entire sections of a page without having to tediously reapply the same sets of styles. This will amplify the site-building potential of pattern overrides, zoomed out mode etc.

A full history of the feature can be found in: https://github.com/WordPress/gutenberg/issues/57537

Syncs the changes from: https://github.com/WordPress/gutenberg/pull/57908 & https://github.com/WordPress/gutenberg/pull/62125

Detailed test instructions can be found on the linked Gutenberg PRs.

Some specific things to test include:

- Registration of block style variations via theme.json partials in `/styles` and theme style variations
- Existing block style variations such as Button's Outline style continue to display correctly on the frontend
- Custom block style variations can be applied in both editors (they won't display properly until the JS changes land)
- Custom block style variations can be applied to nested blocks in any order (frontend displays fine)
- Inner elements and block types can be styled for block style variations and display on the frontend
- Styles applied to individual block instances take precedence on the frontend
- Unit tests pass

```
npm run test:php -- --filter Tests_Block_Supports_WpCreateBlockStyleVariationInstanceName
npm run test:php -- --filter Tests_Block_Supports_WpGetBlockStyleVariationNameFromClass
npm run test:php -- --filter WP_Block_Supports_Block_Style_Variations_Test
npm run test:php -- --filter Tests_Theme_wpThemeJsonResolver
npm run test:php -- --filter Tests_Theme_wpThemeJson
npm run test:php -- --filter Tests_Theme_ThemeDir
```


Trac ticket: https://core.trac.wordpress.org/ticket/61312

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

aaronrobertshaw committed:
Block Styles: Extend block style variations as mechanism for achieving section styling

aaronrobertshaw committed:
Add new test theme to theme dir test

talldan commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,378 @@
+ *
+ * @return string The unique class name.
+ */
+function wp_get_block_style_variation_class_name( $block, $variation ) {
```
Very minor, but looking at it again I wonder if it could be called `wp_create_block_style_variation_class_name` to differentiate it more from `wp_get_block_style_variation_name_from_class`. Or maybe I need to wear my reading glasses more.

I realize that's annoying though as the change would need to be synced back to gutenberg 😄

talldan replied:
Could possibly also add unit tests for a lot of these utility functions.

aaronrobertshaw replied:
> I realize that's annoying though as the change would need to be synced back to gutenberg

There's the (currently) open bug fix in https://github.com/WordPress/gutenberg/pull/62125 so I can make the tweak here and there.

aaronrobertshaw replied:
Done in https://github.com/WordPress/wordpress-develop/pull/6662/commits/c382c4334b3f9827e0523c6a677f05d8ea244c0c and https://github.com/WordPress/wordpress-develop/pull/6662/commits/6b18d60a5e30ec2f8ee58eab8c6f2a445e84da49

talldan commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,378 @@
+
+	$class_name = ".$class_name";
+
+	if ( ! is_admin() ) {
+		remove_filter( 'wp_theme_json_get_style_nodes', 'wp_filter_out_block_nodes' );
+	}
```
An inline comment explaining why this happens could be useful.

aaronrobertshaw replied:
In a recent commit there was an inline comment already added. Do you think it needs expanding?

talldan replied:
Thanks for adding it. It wasn't there when I started reviewing.

Consider this comment resolved 👍

aaronrobertshaw committed:
Bring block style variations hook up to date

A couple of function were out of date with Gutenberg and a new bug fix is also due to land.

talldan commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,418 @@
+	$class_name         = wp_get_block_style_variation_class_name( $parsed_block, $variation );
+	$updated_class_name = $parsed_block['attrs']['className'] . " $class_name";
+	$variation_instance = substr( $class_name, 9 );
+	$class_name         = ".$class_name";
```
Some of the naming here feels a little confusing. Perhaps this line could be a different variable called `$selector` or something similar.

aaronrobertshaw replied:
I'll update the final `$class_name` assignment here to `$selector` if it helps. The first line here though isn't a selector just a class name so I'll leave that bit.

aaronrobertshaw replied:
Done: https://github.com/WordPress/wordpress-develop/pull/6662/commits/c382c4334b3f9827e0523c6a677f05d8ea244c0c

talldan commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,418 @@
+ *
+ * @param array $variations Shared block style variations.
+ *
+ * @return array Block variations data to be merged under styles.blocks
```
```suggestion
 * @return array Block variations data to be merged under `styles.blocks`.
```

aaronrobertshaw committed:
Address feedback

aaronrobertshaw committed:
Add unit tests for variation class name utils

aaronrobertshaw committed:
Improve inline explanation around fix for complex block selectors

aaronrobertshaw committed:
Bring in feedback from Gutenberg PR

aaronrobertshaw committed:
Ensure block style variation styles load after block library

aaronrobertshaw commented:
FYI this backport has been updated to include a bug fix which landed in Gutenberg via https://github.com/WordPress/gutenberg/pull/62125.

ramonjd commented on src/wp-includes/class-wp-theme-json-resolver.php:
```diff
@@ -701,16 +701,46 @@ private static function recursively_iterate_json( $dir ) {
+	 * @return boolean
+	 */
+	private static function style_variation_has_scope( $variation, $scope ) {
+		if ( 'block' === $scope ) {
+				return isset( $variation['blockTypes'] );
+		}
+
+		if ( 'theme' === $scope ) {
+				return ! isset( $variation['blockTypes'] );
+		}
```
```suggestion
		if ( 'block' === $scope ) {
			return isset( $variation['blockTypes'] );
		}

		if ( 'theme' === $scope ) {
			return ! isset( $variation['blockTypes'] );
		}
```

aaronrobertshaw replied:
Not sure what's going on there. I'll try and get it cleaned up.

ramonjd commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -868,6 +871,27 @@ protected static function sanitize( $input, $valid_block_names, $valid_element_n
+		 * inner `blocks`, the overall schema will be generated in multiple passes.
+		 */
+		foreach ( $valid_block_names as $block ) {
+				$schema_settings_blocks[ $block ]           = static::VALID_SETTINGS;
```
As above, is there a double tab here, or am I seeing things? 😄

ramonjd commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1047,12 +1094,21 @@ protected static function get_blocks_metadata() {
+			// Block style variations can be registered through the WP_Block_Styles_Registry as well as block.json.
+			$registered_styles = $style_registry->get_registered_styles_for_block( $block_name );
+			foreach ( $registered_styles as $style ) {
+					$style_selectors[ $style['name'] ] = static::get_block_style_variation_selector( $style['name'], static::$blocks_metadata[ $block_name ]['selector'] );
```
Double tab

ramonjd commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1047,12 +1094,21 @@ protected static function get_blocks_metadata() {
+			}
+
+			if ( ! empty( $style_selectors ) ) {
+					static::$blocks_metadata[ $block_name ]['styleVariations'] = $style_selectors;
```
-->-->

ramonjd commented on tests/phpunit/tests/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,132 @@
+		);
+
+		$theme_json   = WP_Theme_JSON_Resolver::get_theme_data()->get_raw_data();
+		$group_styles = $theme_json['styles']['blocks']['core/group'] ?? array();
```
Just checking `??` is okay in tests 

I think it's still not "officially" approved for use in  Core source files, but tests might be fine.

aaronrobertshaw replied:
I just looked and `??` is used in several files already in core. My understanding is this is ok given the minimum PHP requirements for WordPress now.

aaronrobertshaw replied:
Some of the places that operator is already in use include:

- Interactivity API
- Elements block support
- Dimensions block support
- Plugin install
- Translation controller
- Style engine
- Blocks including Navigation, Image, Calendar, Avatar, Featured Image, Gallery
- A range of phpunit tests

I'm going to mark these operator comments as resolved but happy to address them if all those examples aren't kosher 🙂

ramonjd replied:
This is the way. Thanks for pulling up those examples 👍🏻 

For the record, here's an open ticket that is a discussion about discussing the use of `??` in Core.

https://core.trac.wordpress.org/ticket/58874

noisysocks replied:
Seems silly not to use `??`. It's part of the language. The coding standards as of today don't make a mention of it, meaning it's not expressly allowed or disallowed.

peterwilsoncc replied:
I think it's generally avoided as it can be confusing. The use of [short ternaries](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#ternary-operator) is explicitly ruled out and given null coalescing is very similar I think it's best to avoid it.

tellthemachines replied:
I don't believe it's avoided for being confusing, and I'd hesitate to put it in the short ternary bucket 😅 as it's a very useful operator, allowing us to replace `isset`. Personally I think it makes the code more readable. 

The main reason it's not in the standards is nobody has added it in there yet, as per the [trac ticket discussion](https://core.trac.wordpress.org/ticket/58874). It's sad that we should be blocked from using a helpful language feature due to bureaucratic detail, but there we are 🤷

ramonjd commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,424 @@
+	}
+
+	preg_match_all( '/\bis-style-(?!default)(\S+)\b/', $class_string, $matches );
+	return $matches[1] ?? null;
```
Is `??` now kosher in Core?

noisysocks commented:
Hello! Happy Friday.

Testing first, and then I’ll look at the code.

I followed the testing steps in https://github.com/WordPress/gutenberg/pull/57908 basically to the letter. Thanks for writing them!

I am testing this branch and have the Gutenberg plugin disabled. It’s literally my first time looking at this feature so if I’m doing something wrong just tell me.

When I put this into my functions.php:

```
register_block_style(
	array( 'core/group', 'core/columns' ),
	array(
		'name'       => 'green',
		'label'      => __( 'Green', 'twentytwentythree' ),
		'style_data' => array(
			'color'    => array(
				'background' => '#4f6f52',
				'text'       => '#d2e3c8',
			),
			'blocks'   => array(
				'core/group' => array(
					'color' => array(
						'background' => '#739072',
						'text'       => '#e3eedd',
					),
				),
			),
			'elements' => array(
				'link' => array(
					'color'  => array(
						'text' => '#ead196',
					),
					':hover' => array(
						'color' => array(
							'text' => '#ebd9b4',
						),
					),
				),
			),
		),
	)
);
```

I received this notice:

```
Notice: Function WP_Block_Styles_Registry::register was called incorrectly. Block name must be a string.
```

I worked around this by replacing the `array(...)` with just `’core/group'`.

I noticed that, with that code above, links within a Group block on the frontend correctly received a #ead196 colour but links within a Group block in the editor remained black. It’s a similar story for the #739072 background on nested Group blocks. I could see that in the frontend but not the editor.

I’m not sure what this means:

> Double check whether block style variations with margin styles work correctly when alongside layout global styles.

Do you have an example to test?

When I added a variation by creating a `styles/variation-a.json` file or by setting `styles.blocks.variations` in theme.json, I saw that Variation A came up as an option for Group and Columns blocks but selecting that variation didn’t change any of the block’s colours.

ramonjd commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,424 @@
+ * @return array The parsed block with block style variation classname added.
+ */
+function wp_render_block_style_variation_support_styles( $parsed_block ) {
+	$classes    = $parsed_block['attrs']['className'] ?? null;
```
Same as above `??`

ramonjd commented on src/wp-includes/block-supports/block-style-variations.php:
```diff
@@ -0,0 +1,424 @@
+		$variation_data = $theme_json['styles']['blocks'][ $parsed_block['blockName'] ]['variations'][ $variation ] ?? array();
+
+		if ( ! empty( $variation_data ) ) {
+			break;
```
I'm seeing double tabs all over the place - I'll just ask if it's intentional here, once 😄 

Maybe it's my browser or something

ramonjd reviewed (commented):
I'm still testing, thought I'd leave a bunch of annoying syntax feedback first 😄

aaronrobertshaw commented:
Thanks for testing @noisysocks and @ramonjd 👍 

> I received this notice:

Sorry, it looks like I missed rebasing this PR after the backport for extending block style registration. I'll get it updated shortly.

> I’m not sure what this means:
> 
> > Double check whether block style variations with margin styles work correctly when alongside layout global styles.

By that it meant, if a block style variation specifies margin styles, they don't incorrectly override margin styles set by layout styles e.g. the left/right margins on top level blocks etc. Once I sort out the other feedback, I'll circle back and try and come up with a concrete example.

tellthemachines commented:
> Sorry, it looks like I missed rebasing this PR after the backport for extending block style registration.

Was just coming here to ask for a rebase 😄

aaronrobertshaw committed:
Fix indentation

aaronrobertshaw commented:
Ok, this has been rebased and the indentation sorted. Well I hope the indentation/linting issues are sorted, I don't trust my local phpcs/phpcbf config right now 😅

tellthemachines commented:
I've been testing this mainly by registering style variations in PHP, both with classic and block themes, and it's working as expected for both ✅ 
Also checked that existing style variations, whether core or theme-declared, are still working correctly ✅

aaronrobertshaw commented:
> I noticed that, with that code above, links within a Group block on the frontend correctly received a #ead196 colour but links within a Group block in the editor remained black. It’s a similar story for the #739072 background on nested Group blocks. I could see that in the frontend but not the editor.

Apologies @noisysocks, I messed up the test instructions. The editor side of this feature requires the JS changes in Gutenberg to function properly there.

I'll update the instructions for others that might be testing this one.

noisysocks commented:
Got it, that makes sense. It think the only issue I'm seeing now then is this one:

> When I added a variation by creating a styles/variation-a.json file or by setting styles.blocks.variations in theme.json, I saw that Variation A came up as an option for Group and Columns blocks but selecting that variation didn’t change any of the block’s colours.

Here's a video:

https://github.com/WordPress/wordpress-develop/assets/612155/187ca985-c7fc-45c7-8e9a-8d00b23ce002

aaronrobertshaw commented:
Appreciate the continued patience here @noisysocks 🙇 

The catch with the variation partial used there is that it references css custom properties not present in TwentyTwentyFour. Switching those up to simple hex values might be the best bet for testing. 

I'll go and update the original PR's snippets so they don't catch anyone else out.

ramonjd commented:
I'm running through the scenarios from the Gutenberg PR https://github.com/WordPress/gutenberg/pull/57908 


So far:

✅  blockTypes blocks are being registered correctly and styles applied on frontend (tested using using theme.json and register_block_style)

<img width="268" alt="Screenshot 2024-05-31 at 2 36 53 pm" src="https://github.com/WordPress/wordpress-develop/assets/6458278/ac69c10b-0053-4ad4-8604-d958e6e62764">

✅ existing block style variations on core blocks continue to work

✅ Applying a variation within a block that already has a variation has the correct styles applied

✅ Setting element styles on an individual block instance continues to take precedence

andrewserong reviewed (commented):
This is testing nicely for me on the site frontend, too. All of the items in the "specific things to test" section worked well for me, with nested styles behaving just as I'd expect:

<img width="634" alt="image" src="https://github.com/WordPress/wordpress-develop/assets/14988353/8e0fa0ef-8d2b-4c85-9c30-bfa8f622bd66">

Looking good to me in manual testing!

tellthemachines reviewed (approved):
Code's looking pretty good to me! Thanks for adding comprehensive tests ❤️

ramonjd commented:
Just finished a round of specificity testing, e.g., Top-level, blocks, elements and CSS is being applied as I'd expect. 

LGTM! 

<img width="150" alt="Screenshot 2024-05-31 at 3 26 40 pm" src="https://github.com/WordPress/wordpress-develop/assets/6458278/126e539d-0528-446f-afb5-97d62cdfa995">

noisysocks commented:
> The catch with the variation partial used there is that it references css custom properties not present in TwentyTwentyFour. Switching those up to simple hex values might be the best bet for testing.
> 
> I'll go and update the original PR's snippets so they don't catch anyone else out.

Makes sense 👍 thanks!

I changed the `var()s` in my test code to be hex colours and it works well now.

noisysocks commented:
Committed in r58264.

mukeshpanchal27 commented on src/wp-includes/class-wp-theme-json-resolver.php:
```diff
@@ -701,16 +701,46 @@ private static function recursively_iterate_json( $dir ) {
+	 *
+	 * @param array  $variation Theme.json shaped style variation object.
+	 * @param string $scope     Scope to check e.g. theme, block etc.
+	 *
```
No empty line between "@param" and "@return"

mukeshpanchal27 commented on src/wp-includes/class-wp-theme-json-resolver.php:
```diff
@@ -701,16 +701,46 @@ private static function recursively_iterate_json( $dir ) {
+	 *              can be applied e.g. theme vs block etc.
+	 *
+	 * @param string $scope The scope or type of style variation to retrieve e.g. theme, block etc.
 	 *
```
Same here

mukeshpanchal27 commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1158,16 +1214,18 @@ public function get_settings() {
-	 *                          The options currently supported are 'scope' that makes sure all style are scoped to a
-	 *                          given selector, and root_selector which overwrites and forces a given selector to be
-	 *                          used on the root node.
+	 *                       The options currently supported are:
+	 *                       - 'scope' that makes sure all style are scoped to a given selector
+	 *                       - `root_selector` which overwrites and forces a given selector to be used on the root node
+	 *                       - `skip_root_layout_styles` which omits root layout styles from the generated stylesheet.
```
Need some space to align the document.

mukeshpanchal27 commented:
@noisysocks ☝️

noisysocks commented:
Thanks @mukeshpanchal27, addressed in r58265. Are you interested in fixing `phpcs` so that it automatically catches this kind of stuff? 😀
